buckets/archive/tracker.go

```diff
@@ -0,0 +1,265 @@
+	ch := make(chan powc.JobEvent, 1)
+	if err := t.pgClient.FFS.WatchJobs(ctx, ch, jid); err != nil {
+		return true, fmt.Sprintf("watching current job %s for bucket %s: %s", jid, buckKey, err), nil
+	}
+
+	var aborted bool
+	var abortMsg string
+	var job ffs.Job
+	select {
+	case <-ctx.Done():
+		log.Infof("job %s status watching canceled", jid)
+		return true, "watching cancelled", nil
+	case s, ok := <-ch:
+		if !ok {
+			return true, "powergate closed communication chan", nil
+		}
+		if s.Err != nil {
+			log.Errorf("job %s update: %s", jid, s.Err)
+			aborted = true
+			abortMsg = s.Err.Error()
+		}
+		job = s.Job
+	}
+
+	if !isJobStatusFinal(job.Status) {
+		return true, "no final status yet", nil
+	}
+
+	// Step 2: On success, save Deal data in the underlying Bucket thread. On
+	// failure save the error message. Also update status on Mongo for the archive.
+	if job.Status == ffs.Success {
+		if err := t.saveDealsInArchive(ctx, buckKey, dbID, dbToken, ffsi.FFSToken, bucketRoot); err != nil {
+			return true, fmt.Sprintf("saving deal data in archive: %s", err), nil
+		}
+	}
+	if err := t.updateArchiveStatus(ctx, ffsi, job, aborted, abortMsg); err != nil {
+		return true, fmt.Sprintf("updating archive status: %s", err), nil
+	}
+
+	msg := "reached final status"
+	if aborted {
+		msg = "aborted with reason " + abortMsg
+	}
+
+	return false, msg, nil
+}
+
+// updateArchiveStatus save the last known job status. It also receives an
+// _abort_ flag which indicates that the provided newJobStatus might not be
+// final. For example, if tracking the Job status changes errored by some network
+// condition, we have only the last known Job status. Powergate would continue doing
+// its work until a final status is reached; it only means we aren't sure how this
+// archive really finished.
+// To track for this situation, we use the _aborted_ and _abortMsg_ parameters.
+// An archive with _aborted_ true should eventually be re-queried to understand
+// how it finished (if wanted).
+func (t *Tracker) updateArchiveStatus(ctx context.Context, ffsi *mdb.FFSInstance, job ffs.Job, aborted bool, abortMsg string) error {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+	lastArchive := &ffsi.Archives.Current
+	if lastArchive.JobID != job.ID.String() {
+		for i := range ffsi.Archives.History {
+			if ffsi.Archives.History[i].JobID == job.ID.String() {
+				lastArchive = &ffsi.Archives.History[i]
+				break
+			}
+		}
+	}
+	lastArchive.JobStatus = int(job.Status)
+	lastArchive.Aborted = aborted
+	lastArchive.AbortedMsg = abortMsg
+	lastArchive.FailureMsg = prepareFailureMsg(job)
+	if err := t.colls.FFSInstances.Replace(ctx, ffsi); err != nil {
+		return fmt.Errorf("updating ffs status update instance data: %s", err)
+	}
+	return nil
+}
+
+func (t *Tracker) saveDealsInArchive(ctx context.Context, key string, dbID thread.ID, dbToken thread.Token, ffsToken string, c cid.Cid) error {
+	opts := tdb.WithToken(dbToken)
+	ctx = common.NewSessionContext(ctx, t.internalSession)
```

Comment:
Using the renamed `gatewaySession` (now `internalSession`) to bypass auth whenever the Threads API call gets re-intercepted again.

Fields `dbID` and `dbToken` were persisted when the archive was registered with `.Track(..)`.